packages/fxa-payments-server/src/routes/ProductV2/SubscriptionCreate/index.tsx

```diff
@@ -0,0 +1,363 @@
+    paymentMethodId: paymentMethod.id,
+    stripe,
+    onSuccess,
+    onFailure,
+    onRetry,
+  };
+
+  if (!retryStatus) {
+    // 3a. Attempt to create the subscription.
+    const createSubscriptionResult = await apiCreateSubscriptionWithPaymentMethod(
+      {
+        priceId: selectedPlan.plan_id,
+        paymentMethodId: paymentMethod.id,
+        idempotencyKey,
+      }
+    );
+    return handlePaymentIntent({
+      invoiceId: createSubscriptionResult.latest_invoice.id,
+      paymentIntentStatus:
+        createSubscriptionResult.latest_invoice.payment_intent.status,
+      paymentIntentClientSecret:
+        createSubscriptionResult.latest_invoice.payment_intent.client_secret,
+      ...commonPaymentIntentParams,
+    });
+  } else {
+    // 3b. Retry payment for the subscription invoice created earlier.
+    const { invoiceId } = retryStatus;
+    const retryInvoiceResult = await apiRetryInvoice({
+      invoiceId: retryStatus.invoiceId,
+      paymentMethodId: paymentMethod.id,
+      idempotencyKey,
+    });
+    return handlePaymentIntent({
+      invoiceId,
+      paymentIntentStatus: retryInvoiceResult.payment_intent.status,
+      paymentIntentClientSecret:
+        retryInvoiceResult.payment_intent.client_secret,
+      ...commonPaymentIntentParams,
+    });
+  }
+}
+
+async function handlePaymentIntent({
+  invoiceId,
+  paymentIntentStatus,
+  paymentIntentClientSecret,
+  paymentMethodId,
+  stripe,
+  onSuccess,
+  onFailure,
+  onRetry,
+}: {
+  invoiceId: string;
+  paymentIntentStatus: string;
+  paymentIntentClientSecret: string | null;
+  paymentMethodId: string;
+  stripe: Pick<Stripe, 'confirmCardPayment'>;
+  onFailure: (error: PaymentError) => void;
+  onRetry: (status: RetryStatus) => void;
+  onSuccess: () => void;
+}): Promise<void> {
+  switch (paymentIntentStatus) {
+    case 'succeeded': {
+      return onSuccess();
+    }
+    case 'requires_payment_method': {
+      return onRetry({ invoiceId });
+    }
+    case 'requires_action': {
+      if (!paymentIntentClientSecret) {
+        return onFailure({ type: 'api_error' });
+      }
+      const confirmResult = await stripe.confirmCardPayment(
+        paymentIntentClientSecret,
+        { payment_method: paymentMethodId }
+      );
+      console.log('CONFIRM RESULT', confirmResult);
+      if (confirmResult.error) {
+        return onFailure(confirmResult.error);
+      }
+      if (!confirmResult.paymentIntent) {
+        return onFailure({ type: 'api_error' });
+      }
+      return handlePaymentIntent({
+        invoiceId,
+        paymentIntentStatus: confirmResult.paymentIntent.status,
+        paymentIntentClientSecret: confirmResult.paymentIntent.client_secret,
+        paymentMethodId,
+        stripe,
+        onSuccess,
+        onFailure,
+        onRetry,
+      });
+    }
+    // Other payment_intent.status cases?
+    default: {
+      console.error('Unexpected payment intent status', paymentIntentStatus);
```

Comment:
Debug logging?

Reply:
No, this is an actual error we'll want to see, if it happens. I think this gets picked up by sentry

Reply:
👍 Also just noticed the tests.